### PR TITLE
InstructorAttributes: remove HTML sanitization before saving for role, displayedName #7726

### LINKED ISSUE #7726: InstructorAttributes: remove HTML sanitization before saving for role, displayedName
As part of #4699, sanitization of such fields should be performed during rendering if required.

### PR DESCRIPTION
**Pull request by shaharyar-shamshi**
Fixes #7726
Removal of html sanitization before saving the role and displayedName of instructor.

### CODE REVIEW

**Comment by shaharyar-shamshi** [comment]
Ongoing

**Comment by shaharyar-shamshi** [comment]
@wkurniawan07  sir I have made an necessary changes kindly review it and suggest any further changes if required

**Comment by shaharyar-shamshi** [comment]
@wkurniawan07 Sir it is ready for review

**Comment by shaharyar-shamshi** [comment]
@wkurniawan07 sir it is ready for review kindly have an look

**Comment by shaharyar-shamshi** [comment]
@wkurniawan07  Sir ready for review

**Comment by xpdavid** [comment]
Hi, @shaharyar-shamshi I believe there are some unnecessary changes in your PR. You can try to rebase your changes on top on the current master branch to solve this and also the conflicts.

**Comment by shaharyar-shamshi** [comment]
@wkurniawan07  sir ready for review

**Comment by whipermr5** [comment]
To reviewer: compare with #7746

**Comment by shaharyar-shamshi** [comment]
@xpdavid  sir ready for review kindly have a look

**Comment by xpdavid** [comment]
@shaharyar-shamshi Before I review, let me check with u something.

The issue u are taking is just a small part of a big issue. After this PR get merged, u *must* provide follow up PRs for data migration. Therefore, the task might not be easy as u thought. In addition, the task is not tagged with `d.Contributes`.

**Comment by shaharyar-shamshi** [comment]
@xpdavid no problem sir I will give follow up PRs .

**Comment by shaharyar-shamshi** [comment]
@xpdavid sir after this I will work on removing HTML sanitization for saving name (Instructor attribute)

File: src/main/java/teammates/storage/search/InstructorSearchDocument.java [hunk 1]
```diff
@@ -96,7 +97,7 @@ public static InstructorSearchResultBundle fromResults(Results<ScoredDocument> r
     private static void sortInstructorResultList(List<InstructorAttributes> instructorList) {
 
         instructorList.sort(Comparator.comparing((InstructorAttributes instructor) -> instructor.courseId)
-                .thenComparing(instructor -> instructor.role)
+                .thenComparing(instructor -> SanitizationHelper.desanitizeIfHtmlSanitized(instructor.role))
```
Comment (xpdavid) [inline src/main/java/teammates/storage/search/InstructorSearchDocument.java]: Add `//TODO` to say that you are going to remove this in the future after data migration.

File: src/main/java/teammates/ui/template/CourseEditInstructorPanel.java [hunk 2]
```diff
@@ -34,6 +35,12 @@ public CourseEditInstructorPanel(int instructorToShowIndex, int instructorIndex,
         }
         this.instructor = instructor;
 
+        if (this.instructor != null) {
```
Comment (xpdavid) [inline src/main/java/teammates/ui/template/CourseEditInstructorPanel.java]: Same, add `//TODO`

File: src/main/webapp/jsp/studentCourseDetails.jsp [hunk 3]
```diff
@@ -35,7 +35,9 @@
       <courseDetails:displayDetails id="<%=Const.ParamsNames.INSTRUCTOR_NAME%>">
         <jsp:attribute name="heading">Instructors:</jsp:attribute>
         <jsp:body>
-          <courseDetails:displayInstructors/>
+           <c:forEach items="${data.studentCourseDetailsPanel.instructors}" var="instructor">
```
Comment (xpdavid) [inline src/main/webapp/jsp/studentCourseDetails.jsp]: Why u want to move the `forEach` loop out?

File: src/test/resources/data/StudentCourseDetailsPageUiTest.json [hunk 4]
```diff
@@ -6,6 +6,13 @@
       "isInstructor": true,
       "email": "tmms.test@gmail.tmt",
       "institute": "TEAMMATES Test Institute 1"
+    },
+     "SCDetailsUiT.instr2": {
```
Comment (xpdavid) [inline src/test/resources/data/StudentCourseDetailsPageUiTest.json]: What is the newly added test data for?

File: src/main/java/teammates/ui/template/FeedbackResponseCommentRow.java [hunk 5]
```diff
@@ -259,8 +260,9 @@ private String getEditedAtText(String lastEditorEmail, Instant createdAt, Instan
         }
         boolean isGiverAnonymous = Const.DISPLAYED_NAME_FOR_ANONYMOUS_PARTICIPANT.equals(commentGiverName);
         return "(last edited "
-                + (isGiverAnonymous ? "" : "by " + instructorEmailNameTable.get(lastEditorEmail) + " ")
-                + "at " + TimeHelper.formatDateTimeForSessions(lastEditedAt, sessionTimeZone)
-                + ")";
+                + (isGiverAnonymous
+                    ? ""
+                    : "by " + SanitizationHelper.sanitizeForHtml(instructorEmailNameTable.get(lastEditorEmail)) + " ")
```
Comment (xpdavid) [inline src/main/java/teammates/ui/template/FeedbackResponseCommentRow.java]: Just a note. This is not relevant for this PR. However, it sanitises the **name** of the instructor in the displayed HTML page. Let's just leave it there.

**Comment by xpdavid** [comment]
Hi @shaharyar-shamshi. Please also include the data migration script in this PR.
Here is how thing goes, after the PR is merged, we will create a issue and the project admin will run the data migration against LIVE server. After the script is run, we shall have another PR to remove the code dealing with legacy data.

**Comment by shaharyar-shamshi** [comment]
@xpdavid  okk sir I will

**Comment by shaharyar-shamshi** [comment]
@xpdavid Sir I have removed all extra changes . Actually I was reading about Data Migration Script but can't able to get any good idea . can you guide me about writing Data Migration script So that when I make a further contribution I can work efficiently . I will be thankful to you for this

**Comment by xpdavid** [comment]
@shaharyar-shamshi You can take a look at files under `client/scripts`. e.g. `DataMigrationForSanitizedDataInStudentProfileAttributes.java` to see how thing works.

**Comment by shaharyar-shamshi** [comment]
@xpdavid sir I have made initial attempt to write a script kindly check is logic part correct . after your review I will add other cosmetic changes like adding comment etc . I will be thankful for your review and also give me feedback so that I can write further script correctly.

**Comment by shaharyar-shamshi** [comment]
@xpdavid sir kindly have a look

File: src/client/java/teammates/client/scripts/DataMigrationForSanitizedDataInInstructorAttribute.java [hunk 6]
```diff
@@ -0,0 +1,103 @@
...
+            return false;
+        } else if (s.indexOf("&lt;") >= 0 || s.indexOf("&gt;") >= 0 || s.indexOf("&quot;") >= 0
+                   || s.indexOf("&#x2f;") >= 0 || s.indexOf("&#39;") >= 0 || s.indexOf("&amp;") >= 0) {
+            return true;
+        }
+        return false;
+    }
+
+    private String fixSanitization(String s) {
+        if (isSanitizedString(s)) {
```
Comment (xpdavid) [inline src/client/java/teammates/client/scripts/DataMigrationForSanitizedDataInInstructorAttribute.java]: `SanitizationHelper.desanitizeIfHtmlSanitized()` and done. No need to write `isSanitizedString `

File: src/main/java/teammates/storage/api/InstructorsDb.java [hunk 7]
```diff
@@ -365,4 +367,22 @@ protected InstructorAttributes makeAttributes(Instructor entity) {
         return InstructorAttributes.valueOf(entity);
     }
 
+    public List<InstructorAttributes> getAllInstructor() {
```
Comment (xpdavid) [inline src/main/java/teammates/storage/api/InstructorsDb.java]: Pack it all to a single method and write comment to say `remove this after data migration is done`

Comment (whipermr5) [reply src/main/java/teammates/storage/api/InstructorsDb.java]: Just call ofy() directly from the script; no need to add new code in production just to support the script.

**Comment by shaharyar-shamshi** [comment]
@xpdavid sorry sir for delay actually I am busy due to some academic assignment I will make all the changes soon

**Comment by shaharyar-shamshi** [comment]
@xpdavid sir I have made all the changes . and run the script on local server [this is link of output](https://paste.ubuntu.com/p/SNRndG7b35/)

**Comment by shaharyar-shamshi** [comment]
@xpdavid sir kindly have a look

File: src/client/java/teammates/client/scripts/DataMigrationForSanitizedDataInInstructorAttribute.java [hunk 8]
```diff
@@ -0,0 +1,113 @@
...
+        updateInstructor(instructor);
+    }
+
+    @Override
+    protected void postAction() {
+        // nothing to do
+    }
+
+    private String fixSanitization(String s) {
+        SanitizationHelper.desanitizeIfHtmlSanitized(s);
```
Comment (xpdavid) [inline src/client/java/teammates/client/scripts/DataMigrationForSanitizedDataInInstructorAttribute.java]: You need return the result. String is immutable in Java.

File: src/client/java/teammates/client/scripts/DataMigrationForSanitizedDataInInstructorAttribute.java [hunk 9]
```diff
@@ -0,0 +1,113 @@
...
+                throws InvalidParametersException, EntityDoesNotExistException {
+
+        if (!instructor.isValid()) {
+            throw new InvalidParametersException(instructor.getInvalidityInfo());
+        }
+
+        instructorDb.updateInstructorByEmail(instructor);
+    }
+
+    /**
```
Comment (xpdavid) [inline src/client/java/teammates/client/scripts/DataMigrationForSanitizedDataInInstructorAttribute.java]: No need to have this commet in the script.

File: src/client/java/teammates/client/scripts/DataMigrationForSanitizedDataInInstructorAttribute.java [hunk 10]
```diff
@@ -0,0 +1,113 @@
...
+     * TODO remove after data Migration.
+     */
+    public List<InstructorAttributes> getAllInstructor() {
+        Map<String, InstructorAttributes> result = new LinkedHashMap<>();
+
+        for (InstructorAttributes instructor : getAllCourseInstructor()) {
+            result.put(instructor.getIdentificationString(), instructor);
+        }
+        return new ArrayList<>(result.values());
```
Comment (xpdavid) [inline src/client/java/teammates/client/scripts/DataMigrationForSanitizedDataInInstructorAttribute.java]: Why we need to load it to a HashMap and then only get its value?

File: src/client/java/teammates/client/scripts/DataMigrationForSanitizedDataInInstructorAttribute.java [hunk 11]
```diff
@@ -0,0 +1,113 @@
...
+    }
+
+    @Override
+    protected boolean isPreview() {
+        return true;
+    }
+
+    @Override
+    protected List<InstructorAttributes> getEntities() {
+        return getAllInstructor();
```
Comment (xpdavid) [inline src/client/java/teammates/client/scripts/DataMigrationForSanitizedDataInInstructorAttribute.java]: Can just use `getAllCourseInstructor()`?

File: src/client/java/teammates/client/scripts/DataMigrationForSanitizedDataInInstructorAttribute.java [hunk 12]
```diff
@@ -0,0 +1,113 @@
...
+
+        for (InstructorAttributes instructor : getAllCourseInstructor()) {
+            result.put(instructor.getIdentificationString(), instructor);
+        }
+        return new ArrayList<>(result.values());
+    }
+
+    @Deprecated
+    public List<InstructorAttributes> getAllCourseInstructor() {
+        return instructorDb.makeAttributes(getCourseInstructorEntities());
```
Comment (xpdavid) [inline src/client/java/teammates/client/scripts/DataMigrationForSanitizedDataInInstructorAttribute.java]: You can just use `InstructorAttributes.valueOf(entity);`. No need to touch the production code :)

**Comment by shaharyar-shamshi** [comment]
@xpdavid sir I have made all the changes . Two test cases are failing due to problem as describle in issue #8640 . kindly review the changes

**Comment by shaharyar-shamshi** [comment]
@xpdavid sir kindly have a look

**Comment by shaharyar-shamshi** [comment]
@xpdavid sir kindly have a look . I have made all the changes.

**Comment by shaharyar-shamshi** [comment]
@xpdavid sir kindly have a look on the changes

File: src/main/java/teammates/ui/template/FeedbackResponseCommentRow.java [hunk 13]
```diff
@@ -46,7 +47,7 @@ public FeedbackResponseCommentRow(FeedbackResponseCommentAttributes frc, String
         this.sessionTimeZone = sessionTimeZone;
         this.createdAt = TimeHelper.formatDateTimeForSessions(frc.createdAt, this.sessionTimeZone);
         this.commentText = frc.commentText.getValue();
-        this.commentGiverName = getCommentGiverNameFromEmail(giverDisplay);
+        this.commentGiverName = SanitizationHelper.desanitizeIfHtmlSanitized(getCommentGiverNameFromEmail(giverDisplay));
```
Comment (xpdavid) [inline src/main/java/teammates/ui/template/FeedbackResponseCommentRow.java]: This also need to be removed after data migration right? If so, please add comment?

**Comment by shaharyar-shamshi** [comment]
@xpdavid sir I have added the comment . thanks for your review

**Comment by shaharyar-shamshi** [comment]
@whipermr5 sir kindly have a look

**Comment by shaharyar-shamshi** [comment]
@xpdavid sir till this is reviewed by @whipermr5 sir, can I work on desanitization of other instructors attribute

**Comment by xpdavid** [comment]
@shaharyar-shamshi go ahead :)

**Comment by shaharyar-shamshi** [comment]
@whipermr5 sir kindly review the script .

**Comment by shaharyar-shamshi** [comment]
@whipermr5 sir I have made the changes kindly review it

**Comment by damithc** [comment]
@whipermr5 as usual, please create an issue for me to run the script.

**Comment by whipermr5** [comment]
@shaharyar-shamshi Looks correct! You'll need to set isPreview to false (but don't commit the change) and run it again for it to actually update the entities.

**Comment by shaharyar-shamshi** [comment]
@whipermr5  thanks sir , Sir should I open up the issue for Data Migration ?

**Comment by whipermr5** [comment]
@shaharyar-shamshi No need; I'll open it once the script is completed in your next PR (#8760).

**Pull request merged**